core/src/main/java/org/elasticsearch/common/io/stream/StreamableReader.java

```diff
@@ -20,11 +20,17 @@
 
 import java.io.IOException;
 
+/**
+ * Implementers can be read from {@linkplain StreamInput} by calling their {@link #readFrom(StreamInput)} method.
+ *
+ * It is common for implementers of this interface to declare a <code>public static final</code> instance of themselves named PROTOTYPE so
+ * users can call {@linkplain #readFrom(StreamInput)} on it. It is also fairly typical for readFrom to be implemented a method that just
+ * calls a constructor that takes {@linkplain StreamInput} as a parameter. This allows the fields in the implementer to be
+ * <code>final</code>.
+ */
```

Comment:
The interface design is odd at first but makes much more sense with the explanation.